jobs/MeterInstanceJob.js

```diff
@@ -0,0 +1,145 @@
+    return !plan.metered;
+  }
+
+  static enrichEvent(options) {
+    // Add region , service name and plan sku name of the event
+    const serviceId = _.get(options, 'service.service_guid');
+    const planId = _.get(options, 'service.plan_guid');
+    logger.info(`Enriching the metering event ${serviceId}, plan: ${planId}`);
+    options.service.id = catalog.getServiceName(serviceId);
+    options.service.plan = catalog.getPlanSKUFromPlanGUID(serviceId, planId);
+    options.service = _.omit(options.service, ['service_guid', 'plan_guid']);
+    options.consumer.region = config.metering.region;
+    return options;
+  }
+
+  /* jshint ignore:start */
+  static async sendEvent(event) {
+    try {
+      logger.debug('Metering Event details before enriching:', event);
+      if (this.isServicePlanExcluded(event.spec.options) === true) {
+        await this.updateMeterState(CONST.METER_STATE.EXCLUDED, event);
+        return true;
+      }
+      const enrichedUsageDoc = await this.enrichEvent(_.get(event.spec, 'options'));
+      logger.info('Sending enriched document:', enrichedUsageDoc);
+      const validEvent = await maas.client.sendUsageRecord({
+        usage: [enrichedUsageDoc]
+      });
+      if (validEvent !== undefined) {
+        this.updateMeterState(CONST.METER_STATE.METERED, event);
+      }
+      return true;
+    } catch (err) {
+      logger.error('Error occured while metering:', err);
+      await MeterInstanceJob._logMeteringEvent(err, event);
+      await this.updateMeterState(CONST.METER_STATE.FAILED, event, err);
+      return false;
+    }
+  }
+  /* jshint ignore:end */
+
+  static _logMeteringEvent(err, event) {
+    const now = new Date();
+    const secondsSinceEpoch = Math.round(now.getTime() / 1000);
+    const createSecondsSinceEpoch = Math.round(Date.parse(event.spec.options.timestamp) / 1000);
+    logger.debug(`Event Creation timestamp: ${event.spec.options.timestamp} (${createSecondsSinceEpoch}), Current time: ${now} ${secondsSinceEpoch}`);
+    // Threshold needs to be greater than the metering job frequency
+    const thresholdHours = config.metering.error_threshold_hours;
+    if (secondsSinceEpoch - createSecondsSinceEpoch > thresholdHours * 60 * 60) {
+      logger.debug(`Publishing log event for error: ${err}, for event:`, event);
+      const eventLogger = EventLogInterceptor.getInstance(config.internal.event_type, 'internal');
+      const resp = {
+        statusCode: err.status
+      };
+      const check_res_body = false;
+      return eventLogger.publishAndAuditLogEvent(CONST.URL.METERING_USAGE, CONST.HTTP_METHOD.PUT, event, resp, check_res_body);
+    }
+  }
+
+  static updateMeterState(status, event, err) {
+    logger.debug(`Updating meter state to ${status} for event`, event);
+    let status_obj = {
+      state: status
+    };
+    if (err !== undefined) {
+      status_obj.error = utils.buildErrorJson(err);
+    }
+    return apiServerClient.updateResource({
```

Comment:
better to do this in a retry block here.

Reply:
Added 👍